### ISSUE TITLE
Keep the original command in the list of tab-completion

### ISSUE THREAD

**Issue by Aster89** [qutebrowser/qutebrowser#5177]
@The-Compiler , I'm sorry for the wording of the title, but I miss the terminology (and also some english (and it's also late)). This also means I've not been effective in searching in already existing issues about this topic.

# Foreword #
You already know what you are looking for and you type fast like hell, notheless you can make mistakes.

# Workflow #
- You hit <kbd>:</kbd>, <kbd>o</kbd>, or any other key(-combination) which opens up the completion list;
- you start typing something that you are looking for (in history, commands, whatever) to filter the list, and you do it fast, barely giving your eyes the time to check what you write, maybe because it's something you search for quite frequently;
- without loosing momentum, you press <kbd>Tab</kbd>;
- that's the moment you realize you got something wrong and the completion list has not shrinked down to what you expected;
- you want to recover what you had written before hitting <kbd>Tab</kbd>.

# What stings me #
However <kbd>Shift</kbd>+<kbd>Tab</kbd> loops back to the last completion item, rather than to your hand-typed command.

# Desired (expected, personally) behavior #
<kbd>Tab</kbd> and <kbd>Shift</kbd>+<kbd>Tab</kbd> loop on all the completion items, as well as on the line you edited.

**Comment by The-Compiler** [comment]
Coincidentally, /u/phySi0 [mentioned the same thing](https://www.reddit.com/r/commandline/comments/e935c1/what_are_some_clitui_applications_with_vilike/fcs6ooq?utm_source=share&utm_medium=web2x) in a Reddit post recently:

> When tabbing through history/bookmark options, if I get to the end and tab one last time (or shift tab at the beginning), it cycles instead of deselecting once before cycling. If I've got an `:open -t` going and I type `xlib` — looking for Xlib documentation in my history from my aforementioned crashed sessions whilst doing some X programming (inb4 someone mentions that GUI toolkits exist) — I may not find what I'm looking for in the history and decide to just open `xlib` as a location instead (w/o a TLD, so I'm doing an online search).
>
> Vim itself actually works this way — I just tried it — which is where I was expecting it from. I've always wished my ZSH completion worked the same way (I should fix it), but alas… this is one of those subtle UX things that something either gets right or doesn't.

Makes sense to me to do the same thing for qutebrowser as well. No idea how difficult it'd be to implement, it's been a while since I last looked at completion code. cc @rcorre who probably looked at it a bit more than I did.

**Comment by toofar** [comment]
Would having it as a category at the end be sufficient? It seems like it would for both reports but it would be rather non-obvious. Also probably easy and non-intrusive though.

**Comment by The-Compiler** [comment]
Having a completion item for "go back the original text I entered" seems like a weird thing to see from an user's perspective.

**Comment by rien333** [comment]
Actually, this has bothered me occasionally  as well. I'm not sure if it's the exact same problem as raised in this issue, but I kinda like emacs' behavior were the selected completion and what you type are more or less separate. See this asciicast for what I mean: [![asciicast](https://asciinema.org/a/9rIYFvRc9rc0dlfpTVvoRm9pg.svg)](https://asciinema.org/a/9rIYFvRc9rc0dlfpTVvoRm9pg).


This way, while cycling, if you notice that the completions are not completely what you want, you can still edit your original "query". 

I guess you could implement this in qutebrowser as <kbd>Tab</kbd> just selecting, but not _replacing_, your current query. (and the same for 'completion selecting'-keys other than <kbd>Tab</kbd>)

**Comment by rcorre** [comment]
> Would having it as a category at the end be sufficient? It seems like it would for both reports but it would be rather non-obvious. Also probably easy and non-intrusive though.

I think this is the easiest way to implement. It is pretty simple to add a new category, and you'd get a lot of functionality out of the box (including the ability to disable the category for users who don't want it). I suppose one thing that makes it different from other completion items is that the content would change dynamically with each key press. For every other completion, we just build the model once, then filter it down without ever changing the item contents. Maybe we could just hold a reference to that item and re-set the text, I'm not sure.

> Having a completion item for "go back the original text I entered" seems like a weird thing to see from an user's perspective.

Agreed, though there's also a part of me that wonders if having it visible is actually _more_ obvious. I'm not sure. Difficult usability question :)

I guess I'd lean towards invisible, though that may be a more complex implementation.

> I guess you could implement this in qutebrowser as Tab just selecting a completion, but while not replacing your current query. (and the same for 'completion selecting'-keys other than Tab)

I've thought about this a lot too. I find this kind of behavior really nice in [fzf](https://github.com/junegunn/fzf), which I use for a lot of things. On the other hand, sometimes I tab-complete to a url, then use `kill-word` to trim a bit off the end before hitting enter. I suppose for that case you could use `edit-command`, assuming `edit-command` would operate on the currently _completed_ command (not the original command). That's another user-expectation question, I guess.

In summary, I've got no answers and only more questions :laughing: 
I'll stew on it some more, I definitely think there's a worthwhile feature in this, we just need to figure out the ideal behavior, if there is such a thing.

**Comment by rien333** [comment]
> On the other hand, sometimes I tab-complete to a url, then use kill-word to trim a bit off the end before hitting enter. 

Oh yes, the behavior I proposed definitely breaks editing urls, which I do somewhat more frequently than editing other completition commands. Thus, my proposal might not be a good idea after all.

**Comment by ninewise** [comment]
Some shell-history solves the edit-breaking like this: when you hit `Enter` during completion, the current completion item is selected and executed; when you hit `Ctrl+j` during completion, the current completion item is selected but not executed, so you can edit it before hitting `Enter` to execute it.

**Comment by rien333** [comment]
> Having a completion item for "go back the original text I entered" seems like a weird thing to see from an user's perspective.

This did get me thinking, though.

I'm not sure to what extent this is already possible, but I would find it fairly natural if I could  use <kbd>Ctrl-z</kbd> (i.e. an undo operation) to, well, undo typing, or whatever completion has decided for me. Currently, <kbd>Ctrl-z</kbd> does nothing after doing a completion operation, although it does work for other text operations. 

Relatedly, I can't seem to bind the behavior of <kbd>Ctrl-z</kbd> to other keys. @The-Compiler any idea why `config.bind('<Ctrl-/>', 'fake-key <Ctrl-z>', mode='command')` or even `"<Ctrl-/>" : "<Ctrl-z>"` doesn't work in command mode (these bindings are effective in other modes)?

**Comment by The-Compiler** [comment]
@rien333 Probably because Qt's shortcut handling is independent from how other keys are handled. Have you tried `:fake-key -g`?

FWIW I still think this should be handled the same way vim handles this, i.e. tabbing "over" the completion (in either direction) would reset the text before starting to complete again (and wrapping around).

However, it'd also be nice to see how other vim-like applications/browsers handle this case first.

**Comment by rien333** [comment]
> have you tried `:fake-key -g`?

Didn't know about the `-g` flag, that works fine, thanks!

**Comment by Aster89** [comment]
> when you hit `Ctrl+j` during completion, the current completion item is selected but not executed, so you can edit it before hitting `Enter` to execute it.

Unrelated side question: where can I read about this `Ctrl+j` functionality?

**Comment by ninewise** [comment]
> where can I read about this Ctrl+j functionality?

Search for `isearch-terminators` in the bash man page, for instance.

> However, it'd also be nice to see how other vim-like applications/browsers handle this case first.

vis just allows you to escape completion mode with <kbd>Esc</kbd>.

**Comment by JohnGebbie** [comment]
In Vim you can also press `Ctrl+e` to get back to your original typed text. `:help complete_CTRL-E`

**Comment by The-Compiler** [comment]
Note that we already have `<Ctrl+e>` bound in command mode as a [readline shortcut](https://readline.kablamo.org/emacs.html) to jump to the end of the line.
